### PR TITLE
Use link in UI which returned a relative url but not html_url which contains an absolute url

### LINKED ISSUE #19345: Multi-domain support, ROOT_URL problem, use relative URLs in templates instead of absolute URLs
**At the moment, Gitea users must set ROOT_URL correctly in `app.ini`, otherwise many URLs on the web UI will be wrong, which makes the web UI doesn't work properly (CSRF tokens and cookies can not work across domains)**

The ROOT_URL must be set to the full URL which the users visit, with correct scheme(eg: `https://the-gitea-domain/`) and correct sub path if it is used (eg: `https://the-gitea-domain/my-gitea-path/`).

New code should always use relative URLs for Web UI.

----

Many maintainers have answered the issues about incorrect ROOT_URL again and again.

Using `HTMLURL`(and other absolute URLs) in templates is the root case that Gitea can only serve web under ROOT_URL. 

That's why we have to force users to set ROOT_URL correctly.

If all URLs in template can be refactored to use Relative URL correctly, then we do not need to force users set ROOT_URL correctly again and again.

The ROOT_URL is still needed because for example, some webhooks need the absolute URL for external systems, and Gitea should show absolute URL on the UI for clone link, etc.

Related:
* https://github.com/go-gitea/gitea/issues/14254
* https://github.com/go-gitea/gitea/issues/17912
* https://github.com/go-gitea/gitea/issues/17843
* https://github.com/go-gitea/gitea/issues/19340
* https://github.com/go-gitea/gitea/issues/19963
* https://github.com/go-gitea/gitea/issues/6397 (5 up-votes!)
* https://github.com/go-gitea/gitea/issues/8697  (19 up-votes!)
* https://github.com/go-gitea/gitea/issues/20091
* https://github.com/go-gitea/gitea/issues/20450

Some more related: 
* #20436 , after installation, Gitea may redirect user to an incorrect location.
* #20089 , it may resolve the incorrect redirection problem during installation.

The PR to detect the problem:
* https://github.com/go-gitea/gitea/pull/18971

### PR DESCRIPTION
**Pull request by lunny**
partially fix #19345

This PR add some `Link` methods for different objects. The `Link` methods are not different from `HTMLURL`, they are lack of the absolute URL. And most of UI `HTMLURL` have been replaced to `Link` so that users can visit them from a different domain or IP.

This PR also introduces a new javascript configuration `window.config.reqAppUrl` which is different from `appUrl` which is still an absolute url but the domain has been replaced to the current requested domain.

There are still some tasks left

- [ ] check cookie domains and csrf domains
- [x] the clone http url should be changed to depend on `getCurAbsUrl`
- [ ] package download link should be changed to depend on `getCurAbsUrl`
- [ ] some head properties should depend on `window.config.reqAppUrl` or be kept as before?

### CODE REVIEW

**Comment by silverwind** [comment]
Maybe name it `window.config.currentAppUrl`?

**Comment by wxiaoguang** [comment]
> Maybe name it `window.config.currentAppUrl`? Would be more clear.

It's not worth to add a new field in the `window.config` and make it more complex.

There can be a function like `function currentAppUrl() { return ... + config.appSubUrl; }`, re-use the existing `config.appSubUrl` to make the head scripts simple and clear.

**Comment by lunny** [comment]
> Details

I created a function to do that `getCurAbsUrl(relPath)`

File: templates/base/head_script.tmpl [hunk 1]
```diff
@@ -8,7 +8,9 @@ If you introduce mistakes in it, Gitea JavaScript code wouldn't run correctly.
 	window.config = {
 		appUrl: '{{AppUrl}}',
 		appSubUrl: '{{AppSubUrl}}',
-		currentAppUrl: window.location.protocol + '//' + window.location.host +'{{AppSubUrl}}',
+		getCurAbsUrl: function(relUrl) {
+			return window.location.protocol + '//' + window.location.host +'{{AppSubUrl}}' + relUrl;
+		},
```
Comment (silverwind) [inline templates/base/head_script.tmpl]:
```suggestion
		getCurAbsUrl: (relUrl) => `${window.location.origin}/{{AppSubUrl}}${relUrl}`,
```

**Comment by silverwind** [comment]
Why a function thought? I generally see `window.config` as a data-only object.

**Comment by wxiaoguang** [comment]
> Why a function thought? I generally see `window.config` as a data-only object.

I mean no need to add any new field into `window.config`.

File: templates/base/head_script.tmpl (hunk 1 above)
Comment (lunny) [reply templates/base/head_script.tmpl]: Done.

**Comment by lunny** [comment]
I would like to leave those TODO to next PRs.

File: templates/base/head_script.tmpl [hunk 2]
```diff
@@ -8,6 +8,7 @@ If you introduce mistakes in it, Gitea JavaScript code wouldn't run correctly.
 	window.config = {
 		appUrl: '{{AppUrl}}',
 		appSubUrl: '{{AppSubUrl}}',
+		getCurAbsUrl: (relUrl) => `${window.location.origin}{{AppSubUrl}}${relUrl}`,
```
Comment (silverwind) [inline templates/base/head_script.tmpl]: This is now unused, remove?

Comment (lunny) [reply templates/base/head_script.tmpl]: Now it has been used when copying a comment link.

**Comment by silverwind** [comment]
I'm still not happy with having that function in the config which essentially is just duplicating with `AppSubUrl`. I think it's better you move such a function to `utils.js`, and als make it handle the cases of both missing and present trailing slash in `AppSubUrl`. You should probably also write some test cases for it in `utils.test.js`.

**Review by techknowlogick — CHANGES_REQUESTED** [review]
blocking per @silverwind feedback

**Comment by lunny** [comment]
> I'm still not happy with having that function in the config which essentially is just duplicating with `AppSubUrl`. I think config should just be just plain data (e.g. something that can be serialized as JSON), not functions.
> 
> I think it's better you move such a function to `utils.js`, and als make it handle the cases of both missing and present trailing slash in `AppSubUrl` (unless backend guarantees only one variant). You should probably also write some test cases for it in `utils.test.js`.

Done.

File: web_src/js/utils.js [hunk 3]
```diff
@@ -133,3 +133,9 @@ export function convertImage(blob, mime) {
     }
   });
 }
+
+const {appSubUrl} = window.config;
+
+export function getCurAbsUrl(relUrl) {
+  return `${window.location.origin}${appSubUrl}${relUrl}`;
+}
```
Comment (wxiaoguang) [inline web_src/js/utils.js]: This function does wrong.

* It should behave like a browser:
    * if it gets an absolute URL, just returns it
    * if it gets a relative URL, prepends the origin (but do not touch appSubUrl)

The caller should provide a relative URL to it, the backend should already have prepared the AppSubURL for the path.

Comment (lunny) [reply web_src/js/utils.js]: Done.

File: templates/repo/issue/view_content/context_menu.tmpl [hunk 4]
```diff
@@ -6,11 +6,11 @@
...
 		{{$referenceUrl := ""}}
 		{{if .issue}}
-			{{$referenceUrl = Printf "%s#%s" .ctx.Issue.HTMLURL .item.HashTag}}
+			{{$referenceUrl = Printf "%s#%s" .ctx.Issue.Link .item.HashTag}}
 		{{else}}
-			{{$referenceUrl = Printf "%s/files#%s" .ctx.Issue.HTMLURL .item.HashTag}}
+			{{$referenceUrl = Printf "%s/files#%s" .ctx.Issue.Link .item.HashTag}}
 		{{end}}
-		<div class="item context" data-clipboard-text="{{$referenceUrl}}">{{.ctx.locale.Tr "repo.issues.context.copy_link"}}</div>
+		<div class="item context" data-clipboard-text-type="link" data-clipboard-text="{{AppSubUrl}}{{$referenceUrl}}">{{.ctx.locale.Tr "repo.issues.context.copy_link"}}</div>
```
Comment (silverwind) [inline templates/repo/issue/view_content/context_menu.tmpl]:
```suggestion
		<div class="item context" data-clipboard-text-type="url" data-clipboard-text="{{AppSubUrl}}{{$referenceUrl}}">{{.ctx.locale.Tr "repo.issues.context.copy_link"}}</div>
```

File: web_src/js/features/clipboard.js [hunk 5]
```diff
@@ -50,7 +51,11 @@ export function initGlobalCopyToClipboardListener() {
     // in case <button data-clipboard-text><svg></button>, so we just search
     // up to 3 levels for performance
     for (let i = 0; i < 3 && target; i++) {
-      const text = target.getAttribute('data-clipboard-text') || document.querySelector(target.getAttribute('data-clipboard-target'))?.value;
+      let txt = target.getAttribute('data-clipboard-text');
+      if (txt && target.getAttribute('data-clipboard-text-type') === 'link') {
```
Comment (silverwind) [inline web_src/js/features/clipboard.js]:
```suggestion
      if (txt && target.getAttribute('data-clipboard-text-type') === 'url') {
```

**Comment by silverwind** [comment]
I'm fine with the JS after the suggestions.

**Comment by lunny** [comment]
> I'm fine with the JS after the suggestions.

Done

File: models/activities/action.go [hunk 6]
```diff
@@ -223,6 +223,43 @@ func (a *Action) GetRepoAbsoluteLink() string {
 	return setting.AppURL + url.PathEscape(a.GetRepoUserName()) + "/" + url.PathEscape(a.GetRepoName())
 }
 
+// GetCommentHTMLURL returns link to action comment.
+func (a *Action) GetCommentHTMLURL() string {
```
Comment (silverwind) [inline models/activities/action.go:227]: Can we get rid of the term "HTMLURL"? HTML has nothing to do with URLs, it's confusing.

It's just a absolute URL, right? Why not call it that?

Comment (lunny) [reply models/activities/action.go:227]: To align other places, I think we can rename them all in another PR

Comment (wxiaoguang) [reply models/activities/action.go:227]: Agree to keep it at the moment. There are really a lot of "HTMLURL" in old code. It's not worth  to make them inconsistent by this PR, it's better to change them together in the future.

File: web_src/js/utils.js [hunk 7]
```diff
@@ -133,3 +133,10 @@ export function convertImage(blob, mime) {
     }
   });
 }
+
+export function getCurAbsUrl(relUrl) {
+  if (relUrl.startsWith('http://') || relUrl.startsWith('https://')) {
+    return relUrl;
+  }
+  return `${window.location.origin}${relUrl}`;
```
Comment (delvh) [inline web_src/js/utils.js:141]: won't that make problems if `relUrl` does not start with `/`?

File: web_src/js/utils.js [hunk 8]
```diff
@@ -133,3 +133,10 @@ export function convertImage(blob, mime) {
     }
   });
 }
+
+export function getCurAbsUrl(relUrl) {
```
Comment (delvh) [inline web_src/js/utils.js]: What about renaming it to `toAbsoluteURL(relativeURL)`?

File: models/repo/release.go [hunk 9]
```diff
@@ -131,6 +131,11 @@ func (r *Release) HTMLURL() string {
 	return r.Repo.HTMLURL() + "/releases/tag/" + util.PathEscapeSegments(r.TagName)
 }
 
+// Link the url for a release on the web UI. release must have attributes loaded
```
Comment (delvh) [inline models/repo/release.go]: The difference to `HTMLURL` above is not clear.

File: models/issues/pull.go [hunk 10]
```diff
@@ -713,20 +713,20 @@ func GetPullRequestsByHeadBranch(ctx context.Context, headBranch string, headRep
 	return prs, nil
 }
 
-// GetBaseBranchHTMLURL returns the HTML URL of the base branch
-func (pr *PullRequest) GetBaseBranchHTMLURL() string {
+// GetBaseBranchLink returns the relative HTML URL of the base branch
```
Comment (delvh) [inline models/issues/pull.go]:
```suggestion
// GetBaseBranchLink returns the relative URL of the base branch
```

File: templates/swagger/v1_json.tmpl [hunk 11]
```diff
@@ -18893,6 +18893,10 @@
           "type": "string",
           "x-go-name": "LanguagesURL"
         },
+        "link": {
+          "type": "string",
+          "x-go-name": "Link"
+        },
```
Comment (delvh) [inline templates/swagger/v1_json.tmpl:19104]: Should we really provide this in the API?
We won't get rid of it…
On the other hand, that one should probably be included.

File: models/issues/comment.go [hunk 12]
```diff
@@ -391,21 +391,40 @@ func (c *Comment) HTMLURL() string {
...
+		return ""
+	}
+	return c.link(c.Issue.Link())
+}
+
+func (c *Comment) link(baseURL string) string {
 	if c.Type == CommentTypeCode {
 		if c.ReviewID == 0 {
-			return fmt.Sprintf("%s/files#%s", c.Issue.HTMLURL(), c.HashTag())
+			return fmt.Sprintf("%s/files#%s", baseURL, c.HashTag())
```
Comment (zeripath) [inline models/issues/comment.go]: Use `+` instead of `fmt.Sprintf`

```suggestion
			return baseURL + "/files#" + c.HashTag()
```

File: models/issues/comment.go [hunk 13]
```diff
@@ -391,21 +391,40 @@ func (c *Comment) HTMLURL() string {
 		log.Error("loadRepo(%d): %v", c.Issue.RepoID, err)
 		return ""
 	}
+	return c.link(c.Issue.HTMLURL())
```
Comment (zeripath) [inline models/issues/comment.go]: This should probably just be `c.Issue.HTMLURL() + c.link()` or ` + c.hashLink()` or `+ c.partialLink()`

File: models/issues/comment.go [hunk 14]
```diff
@@ -391,21 +391,40 @@ func (c *Comment) HTMLURL() string {
...
+	}
+	err = c.Issue.LoadRepo(db.DefaultContext)
+	if err != nil { // Silently dropping errors :unamused:
+		log.Error("loadRepo(%d): %v", c.Issue.RepoID, err)
+		return ""
+	}
+	return c.link(c.Issue.Link())
+}
+
+func (c *Comment) link(baseURL string) string {
```
Comment (zeripath) [inline models/issues/comment.go]: See above.

**Comment by lunny** [comment]
@silverwind @zeripath Please review again.

**Review by zeripath — APPROVED** [review]

File: models/activities/notification.go [hunk 15]
```diff
@@ -459,6 +459,22 @@ func (n *Notification) HTMLURL() string {
 	return ""
 }
 
+// Link formats a URL-string to the notification
+func (n *Notification) Link() string {
+	switch n.Source {
+	case NotificationSourceIssue, NotificationSourcePullRequest:
+		if n.Comment != nil {
```
Comment (delvh) [inline models/activities/notification.go:466]: Okay, interesting.
1. I would suggest something like `if n.CommentID > 0`, followed by a `n.LoadComment()`
2. I just noticed that `n.LoadComment()` hasn't been implemented at all for `Notification`, only for `NotificationList`. This is a huge architectural mistake that we will have to correct at some point. 
3. Sounds like a lot of refactoring PRs are still waiting to happen. Let's see when I'll be able to tackle at least some of these problems.

File: models/activities/action.go [hunk 16]
```diff
@@ -223,6 +223,43 @@ func (a *Action) GetRepoAbsoluteLink() string {
...
+	return a.getCommentHTMLURL(db.DefaultContext)
+}
+
+func (a *Action) getCommentHTMLURL(ctx context.Context) string {
+	if a == nil {
+		return "#"
+	}
+	if a.Comment == nil && a.CommentID != 0 {
+		a.Comment, _ = issues_model.GetCommentByID(ctx, a.CommentID)
+	}
```
Comment (delvh) [inline models/activities/action.go]: I recommend extracting this into
```go
func (a *Action) loadComment(ctx context.Context) error {
	if a == nil || a.CommentID == 0 || a.Comment != nil {
		return nil
	}
	a.Comment, err := issues_model.GetCommentByID(ctx, a.CommentID)
	return err
}
```

**Review by delvh — COMMENTED** [review]
As I mentioned already, for all these `func Link` / `func HTMLURL` pairs, the difference is not described in the documentation

File: web_src/js/utils.js (hunk 7 above)
Comment (lunny) [reply web_src/js/utils.js:141]: ${window.location.origin} will always not end with `/` and `relUrl` always start with `/`, so I think it's not a problem.

File: web_src/js/utils.js (hunk 8 above)
Comment (lunny) [reply web_src/js/utils.js]: Changed to `toAbsoluteUrl` because javascript has a different name habit.

File: models/repo/release.go (hunk 9 above)
Comment (lunny) [reply models/repo/release.go]: added `relative`

File: models/issues/pull.go (hunk 10 above)
Comment (lunny) [reply models/issues/pull.go]: done.

File: templates/swagger/v1_json.tmpl (hunk 11 above)
Comment (lunny) [reply templates/swagger/v1_json.tmpl:19104]: Because the struct is also used in Gitea's web UI. So I think it cannot be removed easily. Maybe future PR could fix it.

File: models/activities/notification.go (hunk 15 above)
Comment (lunny) [reply models/activities/notification.go:466]: I think you are right. But this is just a copy of `HTMLURL` with slight changes. So both functions should be changed in future PRs. I will keep this PR doing its job.

File: models/activities/action.go (hunk 16 above)
Comment (lunny) [reply models/activities/action.go]: done.

**Comment by wxiaoguang** [comment]
There are many `class="muted"` added in this PR, which is caused by incorrect conflict resolving and merging.

I  would suggest to review the details again.

**Review by delvh — APPROVED** [review]

File: templates/repo/issue/view_content/comments.tmpl [hunk 17]
```diff
@@ -329,7 +329,7 @@
 					<div class="detail">
 						{{svg "octicon-plus"}}
 						<span class="text grey muted-links">
-							<a href="{{.DependentIssue.HTMLURL}}">
+							<a class="muted" href="{{.DependentIssue.Link}}">
```
Comment (wxiaoguang) [inline templates/repo/issue/view_content/comments.tmpl]: Please check this line and below.

`class="muted"` is an unrelated change.

File: templates/repo/view_file.tmpl [hunk 18]
```diff
@@ -110,10 +110,10 @@
 				</table>
 				<div class="code-line-menu ui vertical pointing menu tippy-target">
 					{{if $.Permission.CanRead $.UnitTypeIssues}}
-						<a class="item ref-in-new-issue" href="{{.RepoLink}}/issues/new?body={{.Repository.HTMLURL}}{{printf "/src/commit/"}}{{PathEscape .CommitID}}/{{PathEscapeSegments .TreePath}}" rel="nofollow noindex">{{.locale.Tr "repo.issues.context.reference_issue"}}</a>
+						<a class="item ref-in-new-issue" href="{{.RepoLink}}/issues/new?body={{.Repository.Link}}{{printf "/src/commit/"}}{{PathEscape .CommitID}}/{{PathEscapeSegments .TreePath}}" rel="nofollow noindex">{{.locale.Tr "repo.issues.context.reference_issue"}}</a>
```
Comment (wxiaoguang) [inline templates/repo/view_file.tmpl]: Here seems incorrect.

When you are referring a file link in an issue, you need the full URL at the moment.

IIRC

`https://try.gitea.io/wxiaoguang/test/src/commit/dc757ee96cf02d80dce4afd41e957aa2a2e341f3/b.go#L1` works

But `/wxiaoguang/test/src/commit/dc757ee96cf02d80dce4afd41e957aa2a2e341f3/b.go#L1` doesn't work in the issue body.

See https://try.gitea.io/wxiaoguang/test/issues/26

Comment (lunny) [reply templates/repo/view_file.tmpl]: I will revert this change and add a comment here.

Comment (lunny) [reply templates/repo/view_file.tmpl]: Done.

File: templates/repo/issue/view_content/comments.tmpl (hunk 17 above)
Comment (lunny) [reply templates/repo/issue/view_content/comments.tmpl]: Done.

**Comment by lunny** [comment]
> blocking per @silverwind feedback

I think @silverwind 's concern has been resolved.

**Comment by wxiaoguang** [comment]
Now it looks good to me.

**Comment by jolheiser** [comment]
I think the outstanding changes were given an approving comment earlier from silverwind as well.
